src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -108,33 +122,143 @@ def __init__(
         max_batch_size: int,
         max_batch_time: int,
     ) -> None:
-        self.__commit = commit
-        self.__max_batch_size = max_batch_size
-        self.__max_batch_time = timedelta(milliseconds=max_batch_time)
-        self.__messages_since_last_commit = 0
-        self.__last_commit = datetime.now()
-        self.__ready_to_commit: MutableMapping[Partition, Position] = {}
-        self.__closed = False
+        self._closed: bool = False
+        self._commit = commit
+        # XXX(iker): these max* variables currently don't behave as maximums,
+        # but rather minimums. We should fix this at some point.
+        self._max_batch_size: int = max_batch_size
+        self._max_batch_time: timedelta = timedelta(milliseconds=max_batch_time)
+
+        self._billing_topic = Topic(settings.KAFKA_OUTCOMES_BILLING)
+        self._billing_producer = self._get_billing_producer()
+        # Especially on incidents generating big backlogs, we must expect a lot
+        # of removals from the beginning of a big FIFO. A dequeue provides O(1)
+        # time on removing items from the beginning; while a regular list takes
+        # O(n).
+        self._ongoing_billing_outcomes: Deque[MetricsBucketBilling] = deque()
+        self._ready_to_commit: MutableMapping[Partition, Position] = {}
+        self._last_commit_time: datetime = datetime.now()
+        self._messages_ready_since_last_commit: int = 0
+
+    def _get_billing_producer(self) -> KafkaProducer:
+        servers = _get_bootstrap_servers(topic=self._billing_topic.name, force_cluster=None)
+        return KafkaProducer(
+            build_kafka_configuration(default_config={}, bootstrap_servers=servers)
+        )
 
     def poll(self) -> None:
+        self._mark_commit_ready()
         if self._should_commit():
             self._bulk_commit()
 
+    def _mark_commit_ready(self):
+        """Removes completed futures at the beginning of
+        ongoing_billing_outcomes and adds them to ready_to_commit.
+
+        If a future has errored, log an error but add to ready_to_commit
+        anyways.
+        """
+        while self._ongoing_billing_outcomes:
+            produce_billing, metrics_msg = self._ongoing_billing_outcomes[0]
+            if not produce_billing.done():
+                break
+
+            if produce_billing.exception():
+                logger.error(
+                    "Async future failed in billing metrics consumer.",
+                    exc_info=produce_billing.exception(),
+                    extra={"offset": metrics_msg.offset},
+                )
+
+            self._ongoing_billing_outcomes.popleft()
+            self._ready_to_commit[metrics_msg.partition] = Position(
+                metrics_msg.next_offset, datetime.now()
+            )
+            self._messages_ready_since_last_commit += 1
+
+    def _should_commit(self) -> bool:
```

Comment:
I have moved these methods to improve the readability, but have not modified them.